src/server-glue/webresource-handler.ts

```diff
@@ -0,0 +1,327 @@
+		.map((f) => sqlNameToODataName(f.fieldName));
+};
+
+const getModifiedFields = (request: uriParser.ODataRequest): string[] => {
+	const modifiedFields = request.modifiedFields;
+
+	if (modifiedFields == null) {
+		return [];
+	}
+
+	const allModifiedFields: string[] = Array.isArray(modifiedFields)
+		? modifiedFields.flatMap((mf) => mf?.fields ?? [])
+		: modifiedFields.fields ?? [];
+
+	return allModifiedFields;
+};
+
+const deleteFiles = async (
+	keysToDelete: string[],
+	webResourceHandler: WebResourceHandler,
+) => {
+	const promises = keysToDelete.map((r) => webResourceHandler.removeFile(r));
+	await Promise.all(promises);
+};
+
+const getCreateWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		'POSTRUN-ERROR': async ({ tx, request }) => {
+			tx?.on('rollback', async () => {
+				const fields = getWebResourceFields(request);
+
+				if (fields.length === 0) {
+					return;
+				}
+
+				const keysToDelete: string[] = fields.map(
+					(f) => request.values[f].href,
+				);
+				await deleteFiles(keysToDelete, webResourceHandler);
+			});
+		},
+	} as sbvrUtils.Hooks;
+};
+
+const isDefined = <T>(x: T | undefined | null): x is T => x != null;
+
+const getWebResourcesHrefs = (webResources?: WebResourcesDbResponse[]) => {
+	const resources = webResources ?? [];
+	const hrefs = resources
+		.flatMap((resource) => Object.values(resource ?? {}))
+		.filter(isDefined)
+		.map((resourceKey) => resourceKey.href);
+	return hrefs;
+};
+
+const getRemoveWebResourceHook = (webResourceHandler: WebResourceHandler) => {
+	return {
+		PRERUN: async (args) => {
+			const { api, request } = args;
+			let fields = getWebResourceFields(request);
+
+			if (fields.length === 0) {
+				return;
+			}
+
+			if (request.method === 'PATCH') {
+				const allFields = getModifiedFields(request).map(sqlNameToODataName);
+				fields = fields.filter((f) => allFields.includes(f));
+			}
+
+			if (fields.length === 0) {
+				return;
+			}
+
+			if (request.method === 'PATCH' && request.odataQuery?.key == null) {
+				throw new errors.BadRequestError(
+					'WebResources can only be updated when providing a resource key.',
+				);
+			}
+
+			const ids = await sbvrUtils.getAffectedIds(args);
+			if (ids.length === 0) {
+				return;
+			}
+
+			const webResources = (await api.get({
+				resource: request.resourceName,
+				passthrough: {
+					tx: args.tx,
+					req: permissions.root,
+				},
+				options: {
+					$select: fields,
+					$filter: {
+						id: {
+							$in: ids,
+						},
+					},
+				},
+			})) as WebResourcesDbResponse[] | undefined;
+
+			request.custom.$pineWebResourcesToDelete =
+				getWebResourcesHrefs(webResources);
+		},
+		POSTRUN: ({ tx, request }) => {
+			tx.on('end', () => {
+				const keysToDelete: string[] =
+					request.custom.$pineWebResourcesToDelete || [];
+				// on purpose does not await for this promise to resolve
+				deleteFiles(keysToDelete, webResourceHandler);
```

Comment:
So if the `pineWebResourcesToDelete` is empty the deleteFiles doesn't need to be called, right? I can see that the deleteFiles won't do anything. I'm curious what is cleaner, calling it with empty array or not calling it at all.

Reply:
@fisehara I added as is because I thought having `keysToDelete` / `$pineWebResourcesToDelete` as a `string[]` was cleaner than a `string[] | null` and then having to always check for the null value. But I can change if you think allowing it to be null is better